### DISCUSSION TITLE
await _table.PullItemsAsync(); hangs the app

### DISCUSSION THREAD

**Discussion by kmostofi** [Azure/azure-mobile-apps#704] [Q&A]
Hi,
I followed the documentation and your tutorial video to set up the backend database. The API response and from the client side, I get a successful response. 

`var connectionString = new UriBuilder { Scheme = "file", Path = "dynamo.sqlite", Query = "?mode=rwc" }.Uri.ToString();
                    store = new OfflineSQLiteStore(connectionString);

                    store.DefineTable<Tin_Members>();

                    optionsDb = new DatasyncClientOptions
                    {
                        OfflineStore = store,
                    };

                    _client = new DatasyncClient(Constants.ServiceUri, optionsDb);

                    await _client.InitializeOfflineStoreAsync();

                    _table = _client.GetOfflineTable<Tbl_Members>();`

And this is the synch method

` public async Task RefreshMembersAsync()
        {
            await InitializeAsync();

            await _table.PushItemsAsync();
          **### _await _table.PullItemsAsync();_**

            return;
        }`

When it hits PullItemAsync , nothing happens and the app hangs.

Please note that model are matching with backend models. 

Could you please advice what I am missing?

Kind regards,
Kamal

**Comment by adrianhall** [comment]
What video?  I haven't done a video! :-)

In any case, there is not enough to go on here.  Some things you need to provide:

1. What does your model look like, on both the server side and client side?
2. What does the controller look like on the server side?
3. If you turn off offline (and just use a remote table), does it work?
4. Turn on logging on the client side - what is going on?

**Reply by kmostofi** [reply]
Hi,
I followed the documentation https://learn.microsoft.com/en-us/azure/developer/mobile-apps/azure-mobile-apps/quickstarts/xamarin-forms/offline?pivots=vs2022-mac 
and watched two vides on you tube https://www.youtube.com/watch?v=7ZcchuKWODQ and https://www.youtube.com/watch?v=FbV4VWf7PfQ

Regarding the models:
Server side:
` public class Tin_Cards : EntityTableData
    {
        public string? CardTitle { get; set; }
        public string? PhotoPath { get; set; }
        public int PhotoOrder { get; set; }
    }`

on Client side the model is as follow:
` public class Tin_Cards : DatasyncClientData, IEquatable<Tin_Cards>
    {
        public string CardTitle { get; set; }
        public string PhotoPath { get; set; }
        public int PhotoOrder { get; set; }
        public bool Equals(Tin_Cards other)
          => other != null && other.Id == Id;
    }`

2- Controller looks like the following on the server side:

` [Route("tables/tin_Cards")]
    public class CardController : TableController<Tin_Cards>
    {
        public CardController(AppDbContext context)
            : base()
        {
            Repository = new EntityTableRepository<Tin_Cards>(context);
        }
    }`

I test the API in Postman and when I run it locally and the API returns the data. After publishing to the azure, I am still able to get the data from server.

3: I turned off offline and tried using remote table only, which I get the following error message:
**One or more errors occurred. (Set JArray values with invalid key value: "version". Int32 array index expected.)**

4: I haven't tried this option yet.

Thank you. :-)

**Reply by adrianhall** [reply]
So - here is the problem.  We expect certain fields (ID, Version, UpdatedAt, Deleted) in your model to have certain forms when translated across the wire.  In this case, Version **MUST** be a String when it comes from the server, but it is stored as a byte[] in the database (a thing that is unique to entity framework, since the database maintains the version field for us).

My suspicion is that the server is not serializing the version to a string.  Instead it is transmitting it as an array of bytes.  This would indicate that you haven't used Services.AddDatasyncControllers() OR that you have a previous Services.AddControllers() with a non-compatible Newtonsoft.JSON setup.

Perhaps you can post the contents of your ConfigureServices() or the service setup section of your Program.cs?  Alternatively, post a link to a github repo with your code and I will take a look.

You might also do a GET of an ID in the table and post the result - I'm thinking the Version field won't be a string as it should be.

**Reply by kmostofi** [reply]
Here is the content of program.cs file

```csharp
var builder = WebApplication.CreateBuilder(args);
var connectionString = @"DefaultConnection"

if (connectionString == null)
{
    throw new ApplicationException("DefaultConnection is not set");
}

builder.Services.AddDbContext<AppDbContext>(options => options.UseSqlServer(connectionString));
builder.Services.AddDatasyncControllers();

builder.Services.AddControllers();

builder.Services.AddEndpointsApiExplorer();
builder.Services.AddSwaggerGen();

var app = builder.Build();
if (app.Environment.IsDevelopment())
{
    app.UseSwagger();
    app.UseSwaggerUI();
}
app.UseHttpsRedirection();
app.UseAuthorization();
app.MapControllers();
app.Run();
```

Here is the result of GET of one record. 
` {
            "cardTitle": "1",
            "photoPath": "https://xxxx.blob.core.windows.net/cards/1.jpg",
            "photoOrder": 1,
            "id": "1",
            "updatedAt": "2021-09-22T10:56:40.267+00:00",
            "version": "AAAAAAABhF0=",
            "deleted": false
        },`

I am afraid the project is not in the Github. :-(

Thanks.

**Reply by adrianhall** [reply]
You have an extra call to `builder.Services.AddControllers()` - remove this.

It's probably not the only thing going on in your application, but the other things are not likely to be problematic.  

See https://github.com/Azure/azure-mobile-apps/blob/main/samples/TodoApp/TodoAppService.NET6/Program.cs for a working version.

**Reply by kmostofi** [reply]
Thanks for the point. I removed the line and still get the same error. I also looked at the link too.

I created a private Repo for API and project.Data . I invited you. 
 Many thanks.

**Reply by adrianhall** [reply]
Thanks - will take a look tomorrow now (I'm tied up rest of today)

**Reply by adrianhall** [reply]
Took a quick look now - there is nothing obvious, but you haven't provided the entire application.  Do you have anything in the client application that is doing Newtonsoft.Json settings updates?

If not, the errors don't make a lot of sense and you will have to delve into the inner workings of your app and the library to see what is going on.

**Reply by kmostofi** [reply]
I have dinged in the error and have this stack flow 

` at Newtonsoft.Json.Linq.JArray.set_Item(Object key, JToken value)\n   at Microsoft.Datasync.Client.Table.RemoteTable.GetJTokenFromResponse(ServiceResponse response)\n   at Microsoft.Datasync.Client.Table.RemoteTable.SendRequestAsync(ServiceRequest request, CancellationToken cancellationToken)\n   at Microsoft.Datasync.Client.Table.RemoteTable.GetNextPageAsync(String query, String requestUri, CancellationToken cancellationToken)\n   at Microsoft.Datasync.Client.Table.RemoteTable`1.<GetNextPageAsync>d__26`1[[Tingle.Data.Models.Tin_Members, Tingle.Data, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null],[Tingle.Data.Models.Tin_Members, Tingle.Data, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].MoveNext()\n   at Microsoft.Datasync.Client.Table.FuncAsyncPageable`1.<AsPages>d__2[[Tingle.Data.Models.Tin_Members, Tingle.Data, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].MoveNext()\n   at Microsoft.Datasync.Client.Table.FuncAsyncPageable`1.<AsPages>d__2[[Tingle.Data.Models.Tin_Members, Tingle.Data, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].System.Threading.Tasks.Sources.IValueTaskSource<System.Boolean>.GetResult(Int16 token)\n   at Microsoft.Datasync.Client.Table.AsyncPageable`1.<GetAsyncEnumerator>d__2[[Tingle.Data.Models.Tin_Members, Tingle.Data, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].MoveNext()\n   at Microsoft.Datasync.Client.Table.AsyncPageable`1.<GetAsyncEnumerator>d__2[[Tingle.Data.Models.Tin_Members, Tingle.Data, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].MoveNext()\n   at Microsoft.Datasync.Client.Table.AsyncPageable`1.<GetAsyncEnumerator>d__2[[Tingle.Data.Models.Tin_Members, Tingle.Data, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].System.Threading.Tasks.Sources.IValueTaskSource<System.Boolean>.GetResult(Int16 token)\n   at System.Linq.AsyncEnumerable.<<ToListAsync>g__Core|424_0>d`1[[Tingle.Data.Models.Tin_Members, Tingle.Data, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].MoveNext() in /_/Ix.NET/Source/System.Linq.Async/System/Linq/Operators/ToList.cs:line 36\n  `

Any ideas?

Thanks.

**Reply by adrianhall** [reply]
Still nope.  You've told me exactly what we knew - the problem is in the deserialization of the response.  It doesn't say why that is happening.

Just had a thought - what does your client side model look like?  It should be inheriting from DatasyncClientData.  It feels like you might have a byte[] Version instead of a string Version in your model.

**Reply by kmostofi** [reply]
This is the client side model for this object:

` public class Tin_Members : DatasyncClientData, IEquatable<Tin_Members>
    {
        public string LastName { get; set; }

        public string FirstName { get; set; }

        public int Gender { get; set; }

        public string AboutMe { get; set; }

        public int Age { get; set; }

        public string Email { get; set; }

        public string Phone { get; set; }

        public int PinCode { get; set; }

        public string City { get; set; }

        public string Country { get; set; }

        public string DOB { get; set; }

        public string Confirmation_Code { get; set; }

        public DateTime Confirmation_Date { get; set; }

        public bool Hidden { get; set; }

        public int Popularity { get; set; }

        public int AccountType { get; set; }

        public bool IsPrivate { get; set; }

        public string RelationTypeItem { get; set; }

        public string FilterByAge { get; set; }

        public string FilterbyRegion { get; set; }

        public int Membership { get; set; }

        public string InstaProfilePhoto { get; set; }

        public string FBName { get; set; }

        public string FBProfilePhoto { get; set; }

        public string TwitterUserName { get; set; }

        public string TwitterProfilePhoto { get; set; }

        public Double FilterByGender { get; set; }

        public string ProfilePhoto { get; set; }

        public bool Equals(Tin_Members other)
            => other != null && other.Id == Id && other.Email == Email && other.PinCode == PinCode;

    }`

Kind regards

**Reply by adrianhall** [reply]
That's very strange.  I'm not going to be able to diagnose this without investing significant debugging time and access to the complete code (neither of which I don't have).  It's something to do with the Newtonsoft.Json deserialization.

**Reply by kmostofi** [reply]
Hi,
I got data with the remote table when I read directly from the remote server. However, when I try to use the offline sync, I get the following error message when it tries to PullItemSync **await _table.PullItemsAsync();.**

`**Set JArray values with invalid key value: "version". Int32 array index expected.**`

`  private async Task InitializeAsync()
        {
            if (_initialized)
            {
                return;
            }

            try
            {
                await _asyncLock.WaitAsync();
                if (_initialized)
                {
                    return;
                }

                if (RemoteToLocalService.StorageDb == null ||
                 RemoteToLocalService.StorageDb.Length == 0) RemoteToLocalService.GetLocalDb();
                try
                {
                    var connectionString = new UriBuilder { Scheme = "file", Path = RemoteToLocalService.StorageDb, Query = "?mode=rwc" }.Uri.ToString();
                    store = new OfflineSQLiteStore(connectionString);

                    store.DefineTable<Tin_Members>();

                    optionsDb = new DatasyncClientOptions
                    {
                        OfflineStore = store,
                        IdGenerator = (table) => $"{table}-{Guid.NewGuid().ToString("D").ToUpperInvariant()}",
                        // HttpPipeline = new HttpMessageHandler[] { new LoggingHandler() },
                    };

                    _client = new DatasyncClient(Constants.ServiceUri, optionsDb);

                    await _client.InitializeOfflineStoreAsync();

                    _table = _client.GetOfflineTable<Tin_Members>();

                    await _table.PushItemsAsync();
                    await _table.PullItemsAsync();

                    _initialized = true;
                }
                catch (Exception ex)
                { throw ex; }

            }
            catch (Exception)
            {
                // Re-throw the exception.
                throw;
            }
            finally
            {
                _asyncLock.Release();
            }
        }`

The sqlite created and also the table exist in the offline database.

Any thought?

Many thanks,
Kamal

**Reply by adrianhall** [reply]
As I've said, there is nothing wrong with your code, so something else must be going on.  I cannot reproduce your error except when I'm explicitly setting Newtonsoft.Json settings.  This being the case, there is nothing else I can do unless you can come up with an explicit xUnit test that shows the error.

**Reply by kmostofi** [reply]
Hi. 
I am sorry to bother you again. After digging into the json, I found the API return a nest object of table (data). I tested the API and the result in postman as follows:

`{
    "items": [
        {
            "lastName": "!=!enc!=!inYXm3x7xxNGXmeDArALuf7w==",
            "firstName": "!=!enc!=!JOf4Fu2xxk2kzakIxKU4/iCQ==",
            "gender": 0,
            "aboutMe": "",
            "age": 25,
            "email": "!=!enc!=!YujjVamvY6pxxVApGgy9OhKKf+3aXLwFY6q7qWGZYdCpw=",
            "phone": "!=!enc!=!igWODLaxx2BNEpDp3WCTwhvg==",
            "pinCode": xxx6,
            "city": "London",
            "country": "Bangladesh",
            "dob": "xxx8",
            "confirmation_Code": "517587",
            "confirmation_Date": "2023-07-16T00:00:00",
            "hidden": true,
            "popularity": 1,
            "accountType": 16,
            "isPrivate": false,
            "relationTypeItem": "0",
            "filterByAge": "18,75",
            "filterbyRegion": "London,United Kingdom : Bangladesh",
            "membership": 1,
            "instaProfilePhoto": "",
            "fbName": "",
            "fbProfilePhoto": "",
            "twitterUserName": "",
            "twitterProfilePhoto": "",
            "filterByGender": 14.0,
            "profilePhoto": "https://xxxxxxxx.blob.core.windows.net/tp27398ffffff.jpg",
            "id": "0000e5f404cdcf4932xa19a2f9vd867f20b1",
            "updatedAt": "2023-07-16T15:10:36.234+00:00",
            "version": "AAAAAAACv4k=",
            "deleted": false
        }
    ],
    "count": null,
    "nextLink": "https://localhost:7017/tables/Members?ZUMO-API-VERSION=3.0.0&$skip=1"
}`

When I run the app, the pullitemasync fails and throws the following error as I mentioned above. 
`Set JArray values with invalid key value: "version". Int32 array index expected.` 

Could the issue be related to the nested Json?

What is your thought?

Kind regards,
Kamal

**Reply by adrianhall** [reply]
No, that's not it.  The list operation is an object which wraps the list of items.  You can see the model in both the server and the client code.

The response from the server is as expected.  You are having a JSON decoding issue for the version - it is not decoding the version correctly.  I do not know why, since the client code sets up the Newtonsoft.JSON settings appropriately.  My suspicion is that you have something else ALSO setting up Newtonsoft.JSON - which is what I've said previously.

**Comment by kmostofi** [comment]
Thanks for the confirmation.
When I test the API, postman returns the nested array but after the publication to Azure, data returns in the root nest.

**Reply by adrianhall** [reply]
Questions:

1. Does the mobile app work when talking to your local API, which has the correct response?
2. What does the response look like when you deploy to Azure?
3. How are you deploying to Azure?  What environment variables are available?  What service is hosting?

**Comment by kmostofi** [comment]
Answer: I use the postman and the local API works nicely. It returns the data in the nested format (Items).

Answer: Response is 200 after deployment to Azure.

- Answer: I use Visual Studio Mac to publish the API to Azure. Mobile App Services and Asp app service plan is running.

Do I need to set API management?

**Reply by adrianhall** [reply]
Nope - API Management is not required.

**Reply by adrianhall** [reply]
"Response is 200" doesn't tell me what the response looks like - just the status code, which is almost irrelevant.  You need to post actual output.

**Comment by kmostofi** [comment]
I posted the response result above from local API,, which is 
`{ "items": [ { "lastName": "!=!enc!=!inYXm3x7xxNGXmeDArALuf7w==", "firstName": "!=!enc!=!JOf4Fu2xxk2kzakIxKU4/iCQ==", "gender": 0, "aboutMe": "", "age": 25, "email": "!=!enc!=!YujjVamvY6pxxVApGgy9OhKKf+3aXLwFY6q7qWGZYdCpw=", "phone": "!=!enc!=!igWODLaxx2BNEpDp3WCTwhvg==", "pinCode": xxx6, "city": "London", "country": "Bangladesh", "dob": "xxx8", "confirmation_Code": "517587", "confirmation_Date": "2023-07-16T00:00:00", "hidden": true, "popularity": 1, "accountType": 16, "isPrivate": false, "relationTypeItem": "0", "filterByAge": "18,75", "filterbyRegion": "London,United Kingdom : Bangladesh", "membership": 1, "instaProfilePhoto": "", "fbName": "", "fbProfilePhoto": "", "twitterUserName": "", "twitterProfilePhoto": "", "filterByGender": 14.0, "profilePhoto": "https://xxxxxxxx.blob.core.windows.net/tp27398ffffff.jpg", "id": "0000e5f404cdcf4932xa19a2f9vd867f20b1", "updatedAt": "2023-07-16T15:10:36.234+00:00", "version": "AAAAAAACv4k=", "deleted": false } ], "count": null, "nextLink": "https://localhost:7017/tables/Members?ZUMO-API-VERSION=3.0.0&$skip=1" }`

and from azure app service, I get the following result:

`[ { "lastName": "!=!enc!=!inYXm3x7xxNGXmeDArALuf7w==", "firstName": "!=!enc!=!JOf4Fu2xxk2kzakIxKU4/iCQ==", "gender": 0, "aboutMe": "", "age": 25, "email": "!=!enc!=!YujjVamvY6pxxVApGgy9OhKKf+3aXLwFY6q7qWGZYdCpw=", "phone": "!=!enc!=!igWODLaxx2BNEpDp3WCTwhvg==", "pinCode": xxx6, "city": "London", "country": "Bangladesh", "dob": "xxx8", "confirmation_Code": "517587", "confirmation_Date": "2023-07-16T00:00:00", "hidden": true, "popularity": 1, "accountType": 16, "isPrivate": false, "relationTypeItem": "0", "filterByAge": "18,75", "filterbyRegion": "London,United Kingdom : Bangladesh", "membership": 1, "instaProfilePhoto": "", "fbName": "", "fbProfilePhoto": "", "twitterUserName": "", "twitterProfilePhoto": "", "filterByGender": 14.0, "profilePhoto": "https://xxxxxxxx.blob.core.windows.net/tp27398ffffff.jpg", "id": "0000e5f404cdcf4932xa19a2f9vd867f20b1", "updatedAt": "2023-07-16T15:10:36.234+00:00", "version": "AAAAAAACv4k=", "deleted": false } ]`

**Reply by adrianhall** [reply]
But that makes no sense - it's telling me that you are not running the same code in App Service vs. locally.  If you were running the same code and the same environment, then the outputs would be the same.
